### PR TITLE
[PhpBridge] add PHPUnit 7 support to SymfonyTestsListener

### PR DESCRIPTION
**Pull request by shieldo**
| Q             | A
| ------------- | ---
| Branch?       | 3.4 up to 4.0 for bug fixes 
| Bug fix?      | yes
| New feature?  | no 
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes
| Fixed tickets | #26017 
| License       | MIT

Add support for the new PHPUnit 7 major release to the PHP Bridge for PHPUnit.

I wasn't sure about making a second legacy class here, or the naming of that class - but it seems like something like this would be required. Can re-name if there's a better suggested approach.

### CODE REVIEW

**Comment by shieldo** [comment]
There's also the question of whether having return types in a file that's being parsed by <PHP 7 will cause an error even if that class isn't to be executed in that runtime. Can anyone comment on that? It might be that this is only appropriate to submit to 4.0+ branches.

**Comment by nicolas-grekas** [comment]
AFAIK, phpunit 7 supports *not* using void, isn't it? If yes, then we should just to the strict minimum to make the bridge work with phpunit 7 *without* using features incompatible with our lower supported PHP version, if possible.

**Comment by shieldo** [comment]
My understanding is that because this class is implementing an interface defined within phpunit/phpunit and that interface defines return types, the return types cannot then be omitted in implementing classes.

I suppose conditionally requiring another file where the class is defined might work better, but that also seems quite kludgy.

**Comment by shieldo** [comment]
Another option might be to use `class_alias` for this rather than just for both of the legacy ones.

**Comment by nicolas-grekas** [comment]
Note that on v4, the bridge support PHP 5.5. That's the special exception (same on 3, where PHP 5.3 is the min for the bridge only, this allows running the latest bridge on most supported versions.)

**Comment by shieldo** [comment]
Note taken. I'll probably rework using `class_alias` to progressively enhance, but any other advice/ suggestions welcome.

**Review by Jean85 — APPROVED** [review]
I agree that an other usage of the `class_alias` is the only way to go for now. I would suggest to decide how long we want to support older, unsupported versions of PHPUnit in this bridge, though...

**Comment by shieldo** [comment]
@Jean85 Was going to rework this PR just now to progressively enhance - i.e. to use `class_alias` for the new class with PHP 5.4+ (if we include traits) syntax in order not to introduce parse errors. Is it OK to rework this PR now you've approved it? It'll be the same classes more or less but in a different order.

**Comment by shieldo** [comment]
@Jean85 I've added another commit to convert this to using `class_alias` in the case of PHPUnit 7/ PHP 7.1, in order to leave the original SymfonyTestsListener class parseable by PHP 5.3 and up. If you'd like me to squash the two commits in this PR, let me know - I didn't want to rewrite the history of what you'd already approved.

**Comment by Jean85** [comment]
@shieldo as you can see, my approval tick is not green, I do not have powers here :smile: 
BTW, contributions are normally squashed automatically afterwards, so you don't have to bother about it.

**Comment by vivekagarwal1984** [comment]
When is this getting merged?

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @shieldo.

File: src/Symfony/Bridge/PhpUnit/SymfonyTestsListenerWithReturnTypes.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * Collects and replays skipped tests.
+ *
+ * (Version of SymfonyTestsListener for PHPUnit 7+, and PHP 7.1+.)
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @final
+ */
+class SymfonyTestsListenerWithReturnTypes implements TestListener
```
Comment (Tobion) [inline src/Symfony/Bridge/PhpUnit/SymfonyTestsListenerWithReturnTypes.php]: shouldn't this be marked internal as people should only interact with SymfonyTestsListener?

Comment (nicolas-grekas) [reply src/Symfony/Bridge/PhpUnit/SymfonyTestsListenerWithReturnTypes.php]: yes! see #26139

**Pull request closed without merging**